File: src/backend/reservation/src/main/java/org/taonaw/reservation/domain/model/reservation/Reservation.java
package org.taonaw.reservation.domain.model.reservation;

import lombok.NonNull;
import org.taonaw.reservation.domain.exception.CanNotCancelReservationException;
import org.taonaw.reservation.domain.exception.CanNotChangeUseTimeException;
import org.taonaw.reservation.domain.exception.CanNotChangeUserInformationException;
import org.taonaw.reservation.domain.model.cancellationfeesetting.CancellationFeeSetting;
import org.taonaw.reservation.domain.model.member.Member;
import org.taonaw.reservation.domain.model.member.MemberId;
import org.taonaw.reservation.domain.model.studio.StudioId;

import java.time.LocalDate;
import java.util.Collection;
import java.util.Objects;

public class Reservation {
    private ReservationId reservationId;
    private StudioId studioId;
    private UseTime useTime;
    private UserInformation userInformation;
    private NumberOfUsers numberOfUsers;
    private PracticeType practiceType;
    private UseEquipments useEquipments;
    private boolean isCanceled;

    private Reservation() { }

    public static Reservation newReservation(
            @NonNull StudioId studioId,
            @NonNull UseTime useTime,
            @NonNull UserInformation userInformation,
            @NonNull NumberOfUsers numberOfUsers,
            @NonNull PracticeType practiceType,
            @NonNull UseEquipments useEquipments) {
        var reservation = new Reservation();
        reservation.reservationId = ReservationId.newId();
        reservation.studioId = studioId;
        reservation.useTime = useTime;
        reservation.userInformation = userInformation;
        reservation.numberOfUsers = numberOfUsers;
        reservation.practiceType = practiceType;
        reservation.useEquipments = useEquipments;
        return reservation;
    }

    public static Reservation reconstruct(
            @NonNull ReservationId reservationId,
            @NonNull StudioId studioId,
            @NonNull UseTime useTime,
            @NonNull UserInformation userInformation,
            @NonNull NumberOfUsers numberOfUsers,
            @NonNull PracticeType practiceType,
            @NonNull UseEquipments useEquipments) {
        var reservation = new Reservation();
        reservation.reservationId = reservationId;
        reservation.studioId = studioId;
        reservation.useTime = useTime;
        reservation.userInformation = userInformation;
        reservation.numberOfUsers = numberOfUsers;
        reservation.practiceType = practiceType;
        reservation.useEquipments = useEquipments;
        return reservation;
    }

    public static Reservation newReservationByMember(
            @NonNull StudioId studioId,
            @NonNull UseTime useTime,
            @NonNull Member member,
            @NonNull NumberOfUsers numberOfUsers,
            @NonNull PracticeType practiceType,
            @NonNull UseEquipments useEquipments) {
        return newReservation(
                studioId,
                useTime,
                member.createUserInformation(),
                numberOfUsers,
                practiceType,
                useEquipments);
    }

    public ReservationId getReservationId() {
        return reservationId;
    }

    public StudioId getStudioId() {
        return studioId;
    }

    public UseTime getUseTime() {
        return useTime;
    }

    public UserInformation getUserInformation() {
        return userInformation;
    }

    public NumberOfUsers getNumberOfUsers() {
        return numberOfUsers;
    }

    public PracticeType getPracticeType() {
        return practiceType;
    }

    public Collection<UseEquipment> getUseEquipments() {
        return useEquipments.getUseEquipments();
    }

    public boolean isCanceled() {
        return isCanceled;
    }

    public boolean isOverlapped(@NonNull Reservation other) {
        return studioId.equals(other.studioId) && useTime.isOverlapped(other.useTime);
    }

    public void changeUseTime(
            @NonNull UseTime useTime,
            @NonNull CancellationFeeSetting cancellationFeeSetting,
            @NonNull LocalDate currentDate) {
        if (this.useTime.equals(useTime)) {
            return;
        }
        if (cancellationFeeSetting.isNotFree(this.useTime, currentDate)) {
            throw new CanNotChangeUseTimeException(reservationId);
        }
        this.useTime = useTime;
    }

    public void changeUserInformation(@NonNull UserInformation userInformation) {
        if (this.userInformation.equals(userInformation)) {
            return;
        }
        if (this.userInformation.isMembersInformation() || userInformation.isMembersInformation()) {
            throw new CanNotChangeUserInformationException(reservationId);
        }
        this.userInformation = userInformation;
    }

    public void changeNumberOfUsers(@NonNull NumberOfUsers numberOfUsers) {
        this.numberOfUsers = numberOfUsers;
    }

    public void changePracticeType(@NonNull PracticeType practiceType) {
        this.practiceType = practiceType;
    }

    public void changeUseEquipments(@NonNull UseEquipments useEquipments) {
        this.useEquipments = useEquipments;
    }

    public void cancel() {
        this.isCanceled = true;
    }

    public void cancelByMember(
            @NonNull MemberId memberId,
            @NonNull CancellationFeeSetting cancellationFeeSetting,
            @NonNull LocalDate currentDate) {

        var exceptionBuilder = new CanNotCancelReservationException.Builder(reservationId);

        if (!userInformation.getMemberId().equals(memberId)) {
            exceptionBuilder.cancelByDifferentMember(memberId);
        }
        if (cancellationFeeSetting.isNotFree(useTime, currentDate)) {
            exceptionBuilder.thereIsCancellationFee();
        }

        exceptionBuilder.throwIfErrorExists();

        this.isCanceled = true;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Reservation that = (Reservation) o;
        return reservationId.equals(that.reservationId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(reservationId);
    }
}

File: src/backend/reservation/src/main/kotlin/org/taonaw/reservation/domain/model/reservation/Reservation.kt
package org.taonaw.reservation.domain.model.reservation

import org.taonaw.reservation.domain.model.equipment.Equipments
import org.taonaw.reservation.domain.model.reservation.TotalUsageFee.TotalUsageFee
import org.taonaw.reservation.domain.model.reservationPolicy.ReservationChangingPolicy
import org.taonaw.reservation.domain.model.reservationPolicy.ReservationPolicy
import org.taonaw.reservation.domain.model.usageFeeSetting.UsageFeeCalculationCondition
import org.taonaw.reservation.domain.model.usageFeeSetting.UsageFeeSetting
import java.time.LocalDateTime

/**
 * 予約
 */
class Reservation private constructor(
        val reservationId: ReservationId,
        val details: ReservationDetails,
        val totalUsageFee: TotalUsageFee) {

    companion object {
        fun create(
                details: ReservationDetails,
                reservationPolicy: ReservationPolicy,
                usageFeeSetting: UsageFeeSetting,
                equipments: Equipments,
                reservedAt: LocalDateTime) : Reservation {

            reservationPolicy.validate(details, reservedAt).throwIfHasErrs("予約内容に不備があります。")

            val usageFee = usageFeeSetting.calculateUsageFee(UsageFeeCalculationCondition.from(details), equipments)

            return Reservation(ReservationId.newId(), details, usageFee)
        }
    }

    fun change(
            changingDetails: ReservationDetails,
            reservationPolicy: ReservationPolicy,
            reservationChangingPolicy: ReservationChangingPolicy,
            usageFeeSetting: UsageFeeSetting,
            equipments: Equipments,
            changedAt: LocalDateTime): Reservation {

        reservationChangingPolicy.validate(details, changingDetails, changedAt)
                .addErr(reservationPolicy.validate(changingDetails, changedAt))
                .throwIfHasErrs("予約内容に不備があります。")

        val changingUsageFee = usageFeeSetting.calculateUsageFee(UsageFeeCalculationCondition.from(changingDetails), equipments)

        return Reservation(reservationId, changingDetails, changingUsageFee)
    }

    fun isDuplicated(other: Reservation): Boolean {
        return this != other &&
                details.studioId == other.details.studioId &&
                details.usageTime.isOverlapping(other.details.usageTime)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Reservation

        if (reservationId != other.reservationId) return false

        return true
    }

    override fun hashCode(): Int {
        return reservationId.hashCode()
    }
}